### PR TITLE
Add autocomplete to Itinerary Generator

### PR DESCRIPTION
**Pull request by emilyzhaoo**
Add autocomplete to input box in Itinerary Generator. 

Allow the users to enter addresses more efficiently. Implement for starting address and event address.

### CODE REVIEW

File: planet/src/main/webapp/places.js [hunk 1]
```diff
@@ -292,6 +285,7 @@ function listResults() {
         icon.classList.add('material-icons');
         icon.classList.add('small');
```
Comment (nabil-dbz) [inline planet/src/main/webapp/places.js]: nit: you can do this in one line: icon.classList.add('material-icons', 'small')

File: planet/src/main/webapp/places.js [hunk 2]
```diff
@@ -342,36 +336,33 @@ function listResults() {
...
     }
-    // Add search keyword to header
+    // Add search keyword to header 
     document.getElementById('greeting').innerHTML = 'Find a place: ' + document.getElementById('pac-input').value;
 }
 
 /** Toggle save icon on click */
 function savePlace(x) {
+    let placeID = $(x).attr('name');
+    let userID = currentUser.uid;
```
Comment (nabil-dbz) [inline planet/src/main/webapp/places.js]: Are you sure that currentUser is not 'null'? (the user is logged in)

File: planet/src/main/webapp/itinerary.html [hunk 3]
```diff
@@ -87,4 +87,23 @@
...
+            autocomplete = new google.maps.places.Autocomplete(startAddress,options); 
+            autocomplete = new google.maps.places.Autocomplete(eventAddress,options);
+            autocomplete.setFields(['address_component']);
+            autocomplete.addListener('place_changed', fillAddress);
+        }
+
+        function fillAddress() {
+            let place = autocomplete.getPlace();
+        }
+    </script>
```
Comment (nabil-dbz) [inline planet/src/main/webapp/itinerary.html]: Why did you add these two functions in the html file?

Comment (emilyzhaoo) [reply planet/src/main/webapp/itinerary.html]: I was calling them from the JS file and I kept getting the error "initAutocomplete does not exist" and I tried a bunch of solutions such as moving the api call in the html file and adding async defer but in the end they didn't work

File: planet/src/main/webapp/places.js (hunk 2 above)
Comment (emilyzhaoo) [reply planet/src/main/webapp/places.js]: When the user is logged in it isn't null

Comment (emilyzhaoo) [reply planet/src/main/webapp/places.js]: We will have to implement something that tells the user they have to be logged in to save places

File: planet/src/main/webapp/itinerary.html (hunk 3 above)
Comment (emilyzhaoo) [reply planet/src/main/webapp/itinerary.html]: So this is the solution I went with

File: planet/src/main/webapp/places.js (hunk 2 above)
Comment (nabil-dbz) [reply planet/src/main/webapp/places.js]: If the user is not logged in, you will get an error for trying to access the property uid of undefined. I suggest you add a condition to check that (currentUser !== null).

Comment (emilyzhaoo) [reply planet/src/main/webapp/places.js]: Will do in next PR!

File: planet/src/main/webapp/itinerary.html [hunk 4]
```diff
@@ -87,4 +87,23 @@
...
     </body>
+    <script>
+        function initAutocomplete() {
+            let startAddress = document.getElementById('starting-address');
+            let eventAddress = document.getElementById('add-event-address');
+
+            let options = {
+                types: ['geocode']
+            };
+            autocomplete = new google.maps.places.Autocomplete(startAddress,options);
```
Comment (EkkiD) [inline planet/src/main/webapp/itinerary.html]: do you need to do any setFields or addListeners on this object?

File: planet/src/main/webapp/itinerary.html (hunk 3 above)
Comment (EkkiD) [reply planet/src/main/webapp/itinerary.html]: Which js file did you have them in?

**Review by EkkiD — CHANGES_REQUESTED** [review]
Add some more info to the PR description with more context: https://google.github.io/eng-practices/review/developer/cl-descriptions.html

File: planet/src/main/webapp/itinerary.html (hunk 3 above)
Comment (emilyzhaoo) [reply planet/src/main/webapp/itinerary.html]: Itinerary.js

Comment (emilyzhaoo) [reply planet/src/main/webapp/itinerary.html]: This is the error: 
Screenshot 2020-07-22 at 11 52 33 AM

File: planet/src/main/webapp/itinerary.html (hunk 4 above)
Comment (emilyzhaoo) [reply planet/src/main/webapp/itinerary.html]: Yes, in line 100 setFields() is used to specify which data to return, which in this case is the address component. Line 101 adds a listener to update the input box when a new search is performed

File: planet/src/main/webapp/itinerary.html (hunk 3 above)
Comment (EkkiD) [reply planet/src/main/webapp/itinerary.html]: itinerary.js is a module. functions in a module are not added to the global namespace by default.

If you add this function to the global namespace in itinerary.js, what is the behaviour? See lines 18-28 of that file for examples

File: planet/src/main/webapp/itinerary.html (hunk 4 above)
Comment (EkkiD) [reply planet/src/main/webapp/itinerary.html]: you're creating the autocomplete object, and them immediately replacing it with a different one. If you mean to setsField and addListener on both of them, you'll need to change this code to use a different variable name.

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: planet/src/main/webapp/js/authentication.js [hunk 5]
```diff
@@ -25,7 +25,7 @@ const firebaseConfig = {
     measurementId: 'G-9W48MVBXLE'
 };
 let currentUser;
-let database;
+let database;
```
Comment (azhang72) [inline planet/src/main/webapp/js/authentication.js]: nit: trailing space

File: planet/src/main/webapp/js/itinerary.js [hunk 6]
```diff
@@ -56,7 +56,7 @@ function handleStartingLocationChange() {
         sessionStorage.setItem('start', document.getElementById('starting-address').value);
     } else {
         alert ('Please update your browser'); 
-    }
+    }
```
Comment (azhang72) [inline planet/src/main/webapp/js/itinerary.js]: nit: trailing space

File: planet/src/main/webapp/places.js [hunk 7]
```diff
@@ -147,6 +147,7 @@ function setSearchByButton() {
 function addPlaceDetails() {
     // For each place, get the icon, name and location.
     places.forEach(function(place) {
+
```
Comment (azhang72) [inline planet/src/main/webapp/places.js]: nit: wandering new line

File: planet/src/main/webapp/places.js [hunk 8]
```diff
@@ -206,7 +200,6 @@ function callback(place, status) {
         }
         placeInfo.push(placeDetails);
     }    
-
```
Comment (azhang72) [inline planet/src/main/webapp/places.js]: nit: remove unrelated changes

File: planet/src/main/webapp/places.js [hunk 9]
```diff
@@ -342,36 +336,33 @@ function listResults() {
         div1.appendChild(div2); 
         element.appendChild(div1);
     }
-    // Add search keyword to header
+    // Add search keyword to header
```
Comment (azhang72) [inline planet/src/main/webapp/places.js]: nit: trailing space

**Review by azhang72 — COMMENTED** [review]
a few nits

File: planet/src/main/webapp/itinerary.html (hunk 3 above)
Comment (zhoua0805) [reply planet/src/main/webapp/itinerary.html]: I got the same error when I was implementing the get place details function (PR #35 ), even after I declared it as a global namespace. So I chose to load the google maps js file synchronously, and call initMap at body onload, which solves the problem

Comment (emilyzhaoo) [reply planet/src/main/webapp/itinerary.html]: Okay! I implemented this in the new commit! thanks

**Review by zhoua0805 — APPROVED** [review]
LGTM

File: planet/src/main/java/com/google/planet/data/ItineraryGenerator.java [hunk 10]
```diff
@@ -48,10 +48,12 @@ public Itinerary generateItinerary(List<Event> events) {
 
         Collections.sort(events, Event.SortByOrder);
         Duration [][] travelTimeGraph = getEventsDistanceMatrix(events);
+
```
Comment (EkkiD) [inline planet/src/main/java/com/google/planet/data/ItineraryGenerator.java]: remove these newlines to keep the diff summary concise

**Review by EkkiD — CHANGES_REQUESTED** [review]

**Review by EkkiD — APPROVED** [review]
LGTM

File: planet/src/main/webapp/js/authentication.js [hunk 11]
```diff
@@ -25,7 +25,11 @@ const firebaseConfig = {
     measurementId: 'G-9W48MVBXLE'
 };
 let currentUser;
+<<<<<<< HEAD
+let database;
+=======
```
Comment (EkkiD) [inline planet/src/main/webapp/js/authentication.js]: oops there's still a merge confict here

File: planet/src/main/webapp/js/itinerary.js [hunk 12]
```diff
@@ -26,9 +26,23 @@ window.handleListOptionChange = handleListOptionChange;
...
+/** Adds autocomplete to input boxes */
+function initAutocomplete() {
+    let startAddress = document.getElementById('starting-address');
+    let eventAddress = document.getElementById('add-event-address');
+
+    let options = {
+        types: ['geocode']
+    };
+
+    new google.maps.places.Autocomplete(startAddress,options);
```
Comment (EkkiD) [inline planet/src/main/webapp/js/itinerary.js]: Consider assigning these to variables or globals so that they can be referenced in the future if needed

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: planet/src/main/webapp/js/authentication.js (hunk 11 above)
Comment (emilyzhaoo) [reply planet/src/main/webapp/js/authentication.js]: Oops! my bad

File: planet/src/main/webapp/js/itinerary.js [hunk 13]
```diff
@@ -26,9 +26,23 @@ window.handleListOptionChange = handleListOptionChange;
...
+/** Adds autocomplete to input boxes */
+function initAutocomplete() {
+    let startAddress = document.getElementById('starting-address');
+    let eventAddress = document.getElementById('add-event-address');
+
+    let options = {
+        types: ['geocode']
+    };
+
+    let autocompleteStart = new google.maps.places.Autocomplete(startAddress,options);
```
Comment (EkkiD) [inline planet/src/main/webapp/js/itinerary.js]: The code here is still confusing, but I guess it's a quirk of the places API.

We're creating an object with `new`, then binding it to a variable. Then we're immediately letting it fall out of scope, removing any reference to the autocomplete object. In theory, the javascript's garbage collector should then delete the object, which would cause autocomplete to stop working. However, I guess that isn't happening, since you've tested this and it works.

More about garbage collection here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Memory_Management

I think the most readable solution would be to define variables on the module scope (by declaring them with `let` outside of any function).
Alternatively, add a comment saying that this is safe because the object doesn't get cleaned up by the garbage collector.

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: planet/src/main/webapp/js/itinerary.js (hunk 13 above)
Comment (emilyzhaoo) [reply planet/src/main/webapp/js/itinerary.js]: Okay! Makes sense

File: planet/src/main/webapp/js/itinerary.js [hunk 14]
```diff
@@ -26,9 +26,26 @@ window.handleListOptionChange = handleListOptionChange;
 window.addEvent = addEvent;
 window.saveEvents = saveEvents;
 window.generateItinerary = generateItinerary;
+window.initAutocomplete = initAutocomplete; 
 
 const database = firebase.database();
 
+// Declare global variables 
+// This is safe because the autocomplete objects do not get cleaned up by the garbage collector
```
Comment (EkkiD) [inline planet/src/main/webapp/js/itinerary.js]: by making this global, we're explicitly preventing them from getting cleaned up. This comment is not necessary.

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: planet/src/main/webapp/js/itinerary.js (hunk 14 above)
Comment (emilyzhaoo) [reply planet/src/main/webapp/js/itinerary.js]: I see! Okay will remove it

**Review by EkkiD — APPROVED** [review]
LGTM
